### PR TITLE
[CIS-218] Events

### PR DESCRIPTION
**Pull request by buh**
In PR is registration and implementation of all events.

1. Events has suffix `Event`
2. Events are grouped into 6 groups with the corresponding prefix:
  - `user`
  - `channel`
  - `message`
  - `member`
  - `reaction`
  - `notification`
3. Events `userStartWatching`/`userStopWatching` and `userStartTyping`/`userStopTyping` parsed as `UserWatchingEvent` and `UserTypingEvent` with a property `isStarted` (and convenient property `isStopped`).
4. Missed tests for `UserUpdatedEvent`. It's hard to get a sample JSON.
5. Most changes are typical.
6. Temporary added missed payloads: `MessagePayload` and `ReactionPayload`.
7. Use filter to hide JSON's:
image

### CODE REVIEW

File: StreamChatClient_v3/Database/DatabaseSession_Tests.swift [hunk 1]
```diff
@@ -24,14 +24,33 @@ final class DatabaseSession_Tests: XCTestCase {
...
         
         // Try to load the saved channel from DB
         var loadedChannel: ChannelModel<DefaultDataTypes>? {
             database.viewContext.loadChannel(cid: channelId)
         }
+        
         AssertAsync.willBeEqual(loadedChannel?.cid, channelId)
+        
+        // Try to load the saved current user from DB
+        if let userId = channelPayload.channel.createdBy?.id {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DatabaseSession_Tests.swift:39]: In theory, you can you `if let` but then you need to ad `else { XCTFail() }` to fail the test properly if the user is `nil`. In practice, you know the content of the payload, so it's OK to force unwrap here.

File: StreamChatClient_v3/Database/DatabaseSession_Tests.swift [hunk 2]
```diff
@@ -24,14 +24,33 @@ final class DatabaseSession_Tests: XCTestCase {
...
         }
         
         // Try to load the saved channel from DB
         var loadedChannel: ChannelModel<DefaultDataTypes>? {
             database.viewContext.loadChannel(cid: channelId)
         }
+        
         AssertAsync.willBeEqual(loadedChannel?.cid, channelId)
+        
+        // Try to load the saved current user from DB
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DatabaseSession_Tests.swift:38]: This is not the "current" user. It's just "any" user. `CurentUser` model was introduced in #349.

File: StreamChatClient_v3/Database/DatabaseSession_Tests.swift [hunk 3]
```diff
@@ -24,14 +24,33 @@ final class DatabaseSession_Tests: XCTestCase {
...
+        if let userId = channelPayload.channel.createdBy?.id {
+            var loadedUser: UserModel<DefaultDataTypes.User>? {
+                database.viewContext.loadUser(id: userId)
+            }
+            
+            AssertAsync.willBeEqual(loadedUser?.id, userId)
+        }
+        
+        // Try to load the saved member from DB
+        if let member = channelPayload.channel.members?.first {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DatabaseSession_Tests.swift:48]: Same here ...

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 4]
```diff
@@ -15,11 +26,16 @@ protocol EventWithPayload: Event {
...
 }
 
+/// A protocol for user owner `Event`s.
+protocol EventWithOwnerPayload: EventWithPayload {
+    var createdByUserId: UserId { get }
+}
+
 /// A protocol for channel `Event`s.
-protocol ChannelEvent: Event {
+protocol EventWithChannelId: Event {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift]: Why this change?

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 5]
```diff
@@ -7,6 +7,17 @@ import Foundation
 /// An `Event` object representing an event in the chat system.
 public protocol Event {}
 
+extension Event {
+    static func parsed<ExtraData: ExtraDataTypes, Value>(keyPath: KeyPath<EventPayload<ExtraData>, Value?>,
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift]: Please add a comment explaining what this magic function does

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 6]
```diff
@@ -15,11 +26,16 @@ protocol EventWithPayload: Event {
 }
 
 /// A protocol for user `Event`s.
-protocol UserEvent: Event {
+protocol EventWithUserPayload: EventWithPayload {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift]: What's wrong with `UserEvent`?

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 7]
```diff
@@ -15,11 +26,16 @@ protocol EventWithPayload: Event {
...
     var userId: UserId { get }
 }
 
+/// A protocol for user owner `Event`s.
+protocol EventWithOwnerPayload: EventWithPayload {
+    var createdByUserId: UserId { get }
+}
+
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift]: I don't understand what this is good for and the comment is not helping.

File: StreamChatClient_v3/WebSocketClient/Events/UserEvents.swift [hunk 8]
```diff
@@ -0,0 +1,92 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public struct UserPresenceChanged<ExtraData: ExtraDataTypes>: EventWithUserPayload {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/UserEvents.swift]: You don't follow the naming you're talking about in the description.

File: StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        XCTAssertEqual(event.cid, ChannelId(type: .messaging, id: "new_channel_7070"))
+        XCTAssertEqual(event.reason, "I don't like you 🤮")
+    }
+    
+    func test_userUnbannedEvent() {
+        let event = try! eventDecoder.decode(from: userUnbannedJSON) as! UserUnbanned<DefaultDataTypes>
+        XCTAssertEqual(event.userId, "broken-waterfall-5")
+        XCTAssertEqual(event.cid, ChannelId(type: .messaging, id: "new_channel_7070"))
+    }
+}
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift]: nit: In general, I think it's better to add the implementation and tests in the same commit :)

File: StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+@testable import StreamChatClient_v3
+import XCTest
+
+class UserEvents_Tests: XCTestCase {
+    let userPresenceJSON = XCTestCase.mockData(fromFile: "UserPresence")
+    let userStartWatchingJSON = XCTestCase.mockData(fromFile: "UserStartWatching")
+    let userStopWatchingJSON = XCTestCase.mockData(fromFile: "UserStopWatching")
+    let userBannedJSON = XCTestCase.mockData(fromFile: "UserBanned")
+    let userUnbannedJSON = XCTestCase.mockData(fromFile: "UserUnbanned")
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift:13]: nit: Why not use it in tests directly, instead of having it here?

File: StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    
+    let eventDecoder = EventDecoder<DefaultDataTypes>()
+    
+    func test_userPresenceEvent() throws {
+        let event = try! eventDecoder.decode(from: userPresenceJSON) as! UserPresenceChanged<DefaultDataTypes>
+        XCTAssertEqual(event.userId, "steep-moon-9")
+        XCTAssertNotNil(event.createdAt)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/UserEvents_Tests.swift]: I already commented about this pattern earlier, it applies to all tests.

File: Example/Sources/CustomChatViewController.swift [hunk 12]
```diff
@@ -66,6 +60,38 @@ class CustomChatViewController: ChatViewController {
...
+    func updateNavigationBar(withChannelImage channelImage: UIImage?) {
+        var items: [UIBarButtonItem] = [membersCountButton]
+        
+        if let channelImage = channelImage {
+            items.append(UIBarButtonItem(image: channelImage, style: .plain, target: nil, action: nil))
+        }
+        
+        if (navigationController?.viewControllers.count ?? 1) > 1 {
+            navigationItem.rightBarButtonItems = items
+        } else {
```
Comment (VojtaStavik) [inline Example/Sources/CustomChatViewController.swift:90]: Why there is a v2 change in a v3 PR? 🤯

File: StreamChatClient_v3/APIClient/Endpoints/MemberPayload_Tests.swift [hunk 13]
```diff
@@ -5,11 +5,8 @@
 @testable import StreamChatClient_v3
 import XCTest
 
-class MemberEndpointPayload_Tests: XCTestCase {
-    let memberJSON: Data = {
-        let url = Bundle(for: MemberEndpointPayload_Tests.self).url(forResource: "Member", withExtension: "json")!
-        return try! Data(contentsOf: url)
-    }()
+final class MemberEndpointPayload_Tests: XCTestCase {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/APIClient/Endpoints/MemberPayload_Tests.swift:8]: Sorry, but I think adding `final` to test classes is just unnecessary noise 🤷‍♂️

File: StreamChatClient_v3/Workers/Background/ChannelEventsHandler.swift [hunk 14]
```diff
@@ -8,7 +8,7 @@ import CoreData
 class ChannelEventsHandler<ExtraData: ExtraDataTypes>: EventHandlerWorker<ExtraData> {
     override func handleNewEvent(event: Event) {
         // TODO: Saving the data must happen in the middleware so we're sure the data is saved before the event is published
-//    if let event = event as? AddedToChannel<ExtraData> {
+//    if let event = event as? NotificationAddedToChannel<ExtraData> {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/Background/ChannelEventsHandler.swift:11]: It's great you also took time to update unsued code 😂

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 15]
```diff
@@ -33,3 +33,8 @@ protocol EventWithChannelId: Event {
 protocol EventWithMessagePayload: EventWithUserPayload, EventWithChannelId {
     var messageId: MessageId { get }
 }
+
+/// A protocol for member `Event`s.
+protocol EventWithMemberPayload: EventWithPayload {
+    var userId: UserId { get }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift:44]: this should probably be `memberId`?

File: StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+final class ReactionEvents_Tests: XCTestCase {
+    let newJSON = XCTestCase.mockData(fromFile: "ReactionNew")
+    let updatedJSON = XCTestCase.mockData(fromFile: "ReactionUpdated")
+    let deletedJSON = XCTestCase.mockData(fromFile: "ReactionDeleted")
+    
+    let eventDecoder = EventDecoder<DefaultDataTypes>()
+    let userId = "broken-waterfall-5"
+    let cid = ChannelId(type: .messaging, id: "general")
+    let messageId = "0e042a9c-d648-4a28-8ed6-dbdb2b7b4779"
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift:16]: Are these the data from the payload? If yes, it should be mentioned somewhere.

File: StreamChatClient_v3/APIClient/Endpoints/MemberPayload.swift [hunk 17]
```diff
@@ -15,7 +15,7 @@ struct MemberPayload<ExtraData: UserExtraData>: Decodable {
         case inviteRejectedAt = "invite_rejected_at"
     }
     
-    let user: UserPayload<ExtraData>
+    let user: UserPayload<ExtraData>?
```
Comment (VojtaStavik) [inline StreamChatClient_v3/APIClient/Endpoints/MemberPayload.swift:18]: How can a user be optional in the member payload?

File: StreamChatClient_v3/Database/DTOs/MemberModelDTO.swift [hunk 18]
```diff
@@ -61,10 +61,12 @@ extension NSManagedObjectContext {
         payload: MemberPayload<ExtraUserData>,
         channelId: ChannelId
     ) throws -> MemberDTO {
-        let dto = MemberDTO.loadOrCreate(id: payload.user.id, channelId: channelId, context: self)
+        let dto = MemberDTO.loadOrCreate(id: payload.user!.id, channelId: channelId, context: self)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DTOs/MemberModelDTO.swift:64]: 😢

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift [hunk 19]
```diff
@@ -19,6 +19,11 @@ protocol EventWithUserPayload: EventWithPayload {
     var userId: UserId { get }
 }
 
+/// A protocol for user owner `Event`s.
+protocol EventWithCurrentUserPayload: EventWithPayload {
+    var currentUserId: UserId { get }
+}
+
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/Event.swift:26]: Sorry, but these protocols deserve much better comments. I have no idea why we have them.

File: StreamChatClient_v3/WebSocketClient/Events/EventType.swift [hunk 20]
```diff
@@ -90,7 +108,20 @@ enum EventType: String, Decodable {
         case .reactionUpdated: return try ReactionUpdatedEvent(from: response)
         case .reactionDeleted: return try ReactionDeletedEvent(from: response)
             
+        case .notificationMessageNew: return try NotificationMessageNewEvent(from: response)
+            
+        case .notificationMarkRead:
+            return response.channel == nil
+                ? try NotificationMarkAllReadEvent(from: response)
+                : try NotificationMarkReadEvent(from: response)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/EventType.swift:116]: It's not your fault but this is 🤮

File: StreamChatClient_v3/Models/MessageRead.swift [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    ///   - unreadMessages: Unread message count
+    public init(user: UserModel<ExtraData>, lastReadDate: Date, unreadMessagesCount: Int) {
+        self.user = user
+        self.lastReadDate = lastReadDate
+        self.unreadMessagesCount = unreadMessagesCount
+    }
+    
+    public static func == (lhs: MessageRead, rhs: MessageRead) -> Bool {
+        lhs.user == rhs.user
+    }
```
[the remark this answers is not in the document]
Comment (buh) [reply StreamChatClient_v3/Models/MessageRead.swift:34]: I'm not sure. I moved it to separate file from `ChannelModel`.

File: StreamChatClient_v3/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift [hunk 22]
```diff
@@ -5,38 +5,37 @@
...
 /// Automatically sends a `TypingStop` event if it hasn't come in a specified time after `TypingStart`.
-class TypingStartCleanupMiddleware<ExtraData: ExtraDataTypes>: EventMiddleware {
+final class TypingStartCleanupMiddleware<ExtraData: ExtraDataTypes>: EventMiddleware {
     /// The maximum time the incoming `typingStart` event is valid before a `typingStop` event is emitted automatically.
     static var incomingTypingStartEventTimeout: TimeInterval { 30 }
     
+    let excludedUserIds: Set<UserId>
+    var typingEventTimeoutTimerControls: [UserId: TimerControl] = [:]
+    var timer: Timer.Type = DefaultTimer.self
+
```
[the remark this answers is not in the document]
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift]: If I left it won't compile because it depends on events.

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift]: We don't have a rule that each commit needs to compile.

File: StreamChatClient_v3/WebSocketClient/Events/ChannelEvents_Tests.swift [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+
+class ChannelEvents_Tests: XCTestCase {
+    let addToChannelJSON = XCTestCase.mockData(fromFile: "AddedToChannel")
+    
+    let eventDecoder = EventDecoder<DefaultDataTypes>()
+    let cid = ChannelId(type: .messaging, id: "general")
+    
+    func test_channelEvents_addToChannel() throws {
+        let addToChannelEvent = try! eventDecoder.decode(from: addToChannelJSON) as! AddedToChannel<DefaultDataTypes>
+        XCTAssertEqual(addToChannelEvent.cid, ChannelId(type: .messaging, id: "new_channel_2766"))
```
[the remark this answers is not in the document]
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/ChannelEvents_Tests.swift]: Yes.

File: StreamChatClient_v3/WebSocketClient/Events/EventPayload_Tests.swift [hunk 24]
```diff
@@ -5,15 +5,14 @@
 @testable import StreamChatClient_v3
 import XCTest
 
+typealias DefaultEventPayload = EventPayload<DefaultDataTypes>
```
[the remark this answers is not in the document]
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload_Tests.swift:8]: It's just in tests.

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload_Tests.swift:8]: I think it brings more confusion instead of clarity. Also, wasn't it you who advocated strongly against using "Default"? 😛

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 4 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift]: The name says one thing, and in fact it does not mean that the channel object will be present.

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 6 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift]: I unified event protocols to make them a little bit more clear for combinations of event payloads.

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 7 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift]: This is the case where we have a user in an event and another user who generates it. For example, a user bans another user.

File: Example/Sources/CustomChatViewController.swift (hunk 12 above)
Comment (buh) [reply Example/Sources/CustomChatViewController.swift:90]: I need to get a sample JSON and add this to v2 Example app. I think it useful to keep it if we need to get more JSON data like that.

File: StreamChatClient_v3/APIClient/Endpoints/MemberPayload_Tests.swift (hunk 13 above)
Comment (buh) [reply StreamChatClient_v3/APIClient/Endpoints/MemberPayload_Tests.swift:8]: I don't think it's a noise. It's a good practice to think about that constantly and do the same in SDK code.

File: StreamChatClient_v3/Workers/Background/ChannelEventsHandler.swift (hunk 14 above)
Comment (buh) [reply StreamChatClient_v3/Workers/Background/ChannelEventsHandler.swift:11]: I think it was automatically with renaming.

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 15 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift:44]: Member doesn't have `memberId`. In `MemberDTO` you do this: `channeldId.rawValue + userId`

File: StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift (hunk 16 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift:16]: It's just a value from JSON to check that it was parsed correctly.

File: StreamChatClient_v3/APIClient/Endpoints/MemberPayload.swift (hunk 17 above)
Comment (buh) [reply StreamChatClient_v3/APIClient/Endpoints/MemberPayload.swift:18]: Unfortunately, it can ¯\_(ツ)_/¯ 
Please check JSON files.

File: StreamChatClient_v3/WebSocketClient/Events/EventType.swift (hunk 20 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/EventType.swift:116]: What's wrong with this?

Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/Events/EventType.swift:116]: The fact that the presence of `channelId` differentiates between two different events. As I said, not your fault :) Just the design of these events is not the best...

File: StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift (hunk 16 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/Events/ReactionEvents_Tests.swift:12]: Can you please add a comment explaining this? It's not obvious.

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 7 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift]: What the benefit of having these kinds of protocols?

File: Example/Sources/CustomChatViewController.swift (hunk 12 above)
Comment (VojtaStavik) [reply Example/Sources/CustomChatViewController.swift]: I see. I think you should've created a separate PR for this change. It's totally our of context.

**Comment by VojtaStavik** [comment]
@buh , it's interesting this PR causes our test coverage to decrease 🤔 Can you maybe check which parts we don't test?

File: StreamChatClient_v3/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift (hunk 22 above)
[the remark this answers is not in the document]
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift]: I think it's important. If someone merged a PR to the master that won't compile then I couldn't start my new task. I had to fix compiling issue first. Is it really not so obvious? 😄

File: StreamChatClient_v3/WebSocketClient/Events/Event.swift (hunk 7 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/Event.swift]: I added comments.

File: StreamChat.xcodeproj/project.pbxproj [hunk 25]
```diff
@@ -2797,7 +2799,7 @@
 				79CDE1DD24B321FE0003BD1D /* CurrentUserDTO.swift in Sources */,
 				797A756824814F0D003CF16D /* BundleExtensions.swift in Sources */,
 				79877A192498E4EE00015F8B /* UserPayload.swift in Sources */,
-				792A4F1B247FE84900EAF71D /* ChannelQueryUpdater.swift in Sources */,
+				792A4F1B247FE84900EAF71D /* ChannelListQueryUpdater.swift in Sources */,
```
Comment (VojtaStavik) [inline StreamChat.xcodeproj/project.pbxproj]: This seems like a leftover from some failed conflict merge. Try to rebase with master again.

File: StreamChat.xcodeproj/project.pbxproj [hunk 26]
```diff
@@ -697,7 +698,6 @@
 		79DDF816249CE38B002F4412 /* MockNetworkURLProtocol.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = MockNetworkURLProtocol.swift; sourceTree = "<group>"; };
 		79DDF817249CE38B002F4412 /* RequestRecorderURLProtocol.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = RequestRecorderURLProtocol.swift; sourceTree = "<group>"; };
 		79DDF818249CE38B002F4412 /* AssertNetworkRequest.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = AssertNetworkRequest.swift; sourceTree = "<group>"; };
-		79E2B83A24CAC6C90024752F /* ChangeAggregator.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = ChangeAggregator.swift; sourceTree = "<group>"; };
 		79E2B83F24CAC8D60024752F /* ChangeAggregator.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = ChangeAggregator.swift; sourceTree = "<group>"; };
```
Comment (VojtaStavik) [inline StreamChat.xcodeproj/project.pbxproj:701]: ... so does this. Rebase should fix this

File: StreamChat.xcodeproj/project.pbxproj [hunk 27]
```diff
@@ -2886,7 +2889,7 @@
 				79280F8424891C6A00CDEB89 /* VirtualTime.swift in Sources */,
 				799C9476247D7E94001F1104 /* TemporaryData.swift in Sources */,
 				792921C724C047DD00116BBB /* APIClient_Mock.swift in Sources */,
-				794927F2249E3DE8009D7EB7 /* Event_Tests.swift in Sources */,
+				794927F2249E3DE8009D7EB7 /* EventPayload_Tests.swift in Sources */,
```
Comment (VojtaStavik) [inline StreamChat.xcodeproj/project.pbxproj:2892]: ... and this

File: StreamChatClient_v3/Database/DTOs/CurrentUserDTO_Tests.swift [hunk 28]
```diff
@@ -15,6 +15,7 @@ class CurrentUserModelDTO_Tests: XCTestCase {
     
     func test_currentUserPayload_isStoredAndLoadedFromDB() {
         let userId = UUID().uuidString
+        let extraData = NameAndImageExtraData(name: "Luke", imageURL: URL(string: UUID().uuidString))
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DTOs/CurrentUserDTO_Tests.swift:18]: 🤷‍♂️

File: StreamChatClient_v3/Database/DatabaseSession.swift [hunk 29]
```diff
@@ -66,7 +66,12 @@ extension DatabaseSession {
         }
         
         if let currentUserPayload = payload.currentUser {
-            try saveCurrentUser(payload: currentUserPayload)
+            let currentUserDTO = try saveCurrentUser(payload: currentUserPayload)
+            
+            if let unreadCount = payload.unreadCount {
+                currentUserDTO.unreadChannelsCount = Int16(unreadCount.channels)
+                currentUserDTO.unreadMessagesCount = Int16(unreadCount.messages)
+            }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Database/DatabaseSession.swift:74]: Interesting solution! 👍

File: StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift [hunk 30]
```diff
@@ -100,14 +88,33 @@ struct EventPayload<ExtraData: ExtraDataTypes>: Decodable {
...
+        cid = try container.decodeIfPresent(ChannelId.self, forKey: .cid)
+        currentUser = try container.decodeIfPresent(CurrentUserPayload<ExtraData.User>.self, forKey: .currentUser)
+        user = try container.decodeIfPresent(UserPayload<ExtraData.User>.self, forKey: .user)
+        createdBy = try container.decodeIfPresent(UserPayload<ExtraData.User>.self, forKey: .createdBy)
+        memberContainer = try container.decodeIfPresent(MemberContainerPayload<ExtraData.User>.self, forKey: .memberContainer)
+        channel = try container.decodeIfPresent(ChannelDetailPayload<ExtraData>.self, forKey: .channel)
+        message = try container.decodeIfPresent(MessagePayload<ExtraData>.self, forKey: .message)
+        reaction = try container.decodeIfPresent(ReactionPayload.self, forKey: .reaction)
+        watcherCount = try container.decodeIfPresent(Int.self, forKey: .watcherCount)
+        unreadCount = try? UnreadCount(from: decoder)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift:111]: Will this report parsing error correctly when `UnreadCount` will have an incorrect JSON format?

File: StreamChat.xcodeproj/project.pbxproj (hunk 25 above)
Comment (buh) [reply StreamChat.xcodeproj/project.pbxproj]: How it should be?

File: StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift (hunk 30 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift:111]: Yes, if it present, but incorrect.

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift:111]: Actually no, but I don't know how to implement it optionally

Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift:111]: Let's leave it

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/Events/EventPayload.swift:111]: I found that unread count is always presented in events at least with sample JSON data that I have in tests.
